fili-core/src/test/groovy/com/yahoo/bard/webservice/async/AsyncFunctionalSpec.groovy

```diff
@@ -0,0 +1,211 @@
+     *
+     * @return The map of query parameters for each target.
+     *
+     */
+    abstract Map<String, Closure<Map<String, List<String>>>> getQueryParameters()
+
+    /**
+     * Returns the classes for the Jetty resources (i.e. servlets) that need to be set up in the test environment.
+     *
+     * @return An array of classes representing Jetty resources
+     */
+    Class<?>[] getResourceClasses() {
+        return [DataServlet.class, JobsServlet.class]
+    }
+
+    /**
+     * Returns the generator for the Druid response that should be returned by the backend
+     *
+     * @return The closure that generates the Druid response that should be returned by the test framework's Druid stub
+     */
+    abstract Closure<String> getFakeDruidResponse()
+
+    /**
+     * Returns the status code that the fake Druid should send when sending results back.
+     *
+     * @return The status code that the fake Druid should use when sending results back, defaults to 200 (OK)
+     */
+    int getDruidStatusCode() {
+        return 200
+    }
+
+    def setup() {
+        // Create the test web container to test the resources
+        jtb = buildTestBinder()
+
+        populatePhysicalTableAvailability()
+    }
+
+    /**
+     * Populates the interval availability of the physical tables.
+     * <p>
+     * By default, every Physical table believes that it has complete data from January 1st 2010 to December 31 2500.
+     */
+    void populatePhysicalTableAvailability() {
+        AvailabilityTestingUtils.populatePhysicalTableCacheIntervals(jtb, new Interval("2010-01-01/2500-12-31"))
+    }
+
+    def cleanup() {
+        // Release the test web container
+        jtb.tearDown()
+    }
+
+    JerseyTestBinder buildTestBinder() {
+        new JerseyTestBinder(getResourceClasses())
+    }
+
+    void validateJson(String json) {
+        MAPPER.readTree(json)
+    }
+
+    @Timeout(10)
+    def "The asynchronous workflow executes correctly"() {
+        given: "Initialize the backend with the fake Druid response"
+        injectDruidResponse(getFakeDruidResponse())
+
+        and: "The map of all previous responses"
+        Map<String, Response> previousResponses = [:]
+
+        and: "The closures that perform verification against the results"
+        Map<String, Closure<Void>> resultValidations = getResultAssertions()
+
+        and: "The query parameters for each request"
+        Map<String, Closure<Map<String, List<String>>>> queryParameters = getQueryParameters()
+
+        expect: "All the asynchronous interactions behave appropriately"
+        getResultsToTargetFunctions().each {interactionName, resultsToTarget ->
+            Response response = makeAbstractRequest(
+                    resultsToTarget(previousResponses),
+                    queryParameters.get(interactionName)(previousResponses)
+            )
+            response.bufferEntity()
+            resultValidations[interactionName](response)
```

Comment:
Also, we should use 1 mapp access style: either the index accessor as used on this line, or the getter as used 3 lines up. I like index access, personally, but we should use just one.